File: src/main/java/com/github/nsc/de/compiler/lexer/token/Token.java
package com.github.nsc.de.compiler.lexer.token;

import com.github.nsc.de.compiler.lexer.Position;

import java.util.Objects;

public class Token {

    private final TokenType type;
    private final String value;
    private final Position start;
    private final Position end;

    public Token(TokenType type, String value, Position start, Position end) {
        this.type = type;
        this.value = value;
        this.start = start;
        this.end = end;
    }

    public Token(TokenType type, String value, Position start) { this(type, value, start, start); }
    public Token(TokenType type, Position start, Position end) { this(type, null, start, end); }
    public Token(TokenType type, Position start) { this(type, null, start, start); }

    public TokenType getType() { return type; }
    public String getValue() { return value; }
    public Position getStart() { return start; }
    public Position getEnd() { return end; }
    public int getBefore() { return start.getIndex(); }
    public int getAfter() { return end.getIndex() + 1; }

    @Override
    public String toString() {
        return start == end ? (value != null ? "" +
                "Token{" + "type=" + type + ", value=" + value + ", position=" + start + '}' :
                "Token{" + "type=" + type +  ", position=" + start + '}')
              : (value != null ? "" +
                "Token{" + "type=" + type + ", value=" + value + ", start=" + start + ", end=" + end + '}' :
                "Token{" + "type=" + type +  ", position=" + start + ", end=" + end + '}');
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Token token = (Token) o;
        return type == token.type &&
                Objects.equals(value, token.value);
    }

    @Override
    public int hashCode() {
        return Objects.hash(type, value);
    }
}

File: lexer/src/commonMain/kotlin/com/github/nsc/de/shake/lexer/token/Token.kt
package com.github.nsc.de.shake.lexer.token


/**
 * The input of the [com.github.nsc.de.shake.lexer.Lexer] gets converted into [Token]s. These get parsed
 * by the parser
 *
 * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
 */
@Suppress("unused")
class Token
/**
 * Constructor for [Token]
 *
 * @param type the [Token.type] of the [Token]
 * @param value the [Token.value] of the [Token]
 * @param start the [Token.start] of the [Token]
 * @param end the [Token.end] of the [Token]
 *
 * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
 *
 * @see Token
 * @see Token.type
 * @see Token.value
 * @see Token.start
 * @see Token.end
 *
 */(
    /**
     * The type of the [Token]
     *
     * @see Token
     */
    val type: Byte,
    /**
     * The value of the [Token] (This is for identifiers, strings or numbers. If not necessary this is null)
     *
     * @see Token
     */
    val value: String?,
    /**
     * The starting [com.github.nsc.de.shake.util.characterinput.position.Position] of the [Token]
     *
     * @see Token
     */
    val start: Int,
    /**
     * The ending [com.github.nsc.de.shake.util.characterinput.position.Position] of the [Token]
     *
     * @see Token
     */
    val end: Int
) {

    /**
     * Constructor for [Token]
     *
     * @param type the [type] of the [Token]
     * @param value the [value] of the [Token]
     * @param position the [start] and [end] and of the [Token]
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     *
     * @see Token
     * @see type
     * @see value
     * @see start
     * @see end
     */
    constructor(type: Byte, value: String?, position: Int) : this(type, value, position, position)

    /**
     * Constructor for [Token]
     *
     * @param type the [type] of the [Token]
     * @param start the [start] of the [Token]
     * @param end the [end] of the [Token]
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     *
     * @see Token
     * @see type
     * @see start
     * @see end
     */
    constructor(type: Byte, start: Int, end: Int) : this(type, null, start, end)

    /**
     * Constructor for [Token]
     *
     * @param type the [type] of the [Token]
     * @param position the [start] and [end] and of the [Token]
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     *
     * @see Token
     * @see type
     * @see start
     * @see end
     */
    constructor(type: Byte, position: Int) : this(type, null, position, position)

    override fun toString(): String {
        return if (start == end) if (value != null) "" +
                "Token{" + "type=" + type + ", value=" + value + ", position=" + start + '}' else "Token{type=$type, position=$start}" else if (value != null) "" +
                "Token{" + "type=" + type + ", value=" + value + ", start=" + start + ", end=" + end + '}' else "Token{type=$type, position=$start, end=$end}"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if(other is Byte) return other == this.type
        if (other == null || other !is Token) return false
        return type == other.type &&
                value == other.value
    }

    override fun hashCode(): Int = hashAll(type, value)

    private fun hashAll(vararg vals: Any?): Int {
        var res = 0
        for (v in vals) {
            res += v.hashCode()
            res *= 31
        }
        return res
    }

}